Dial in test

damencho opened the pull request:
Depends on https://github.com/jitsi/jitsi-meet/pull/3778

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/DialInAudioTest.java:
```diff
@@ -0,0 +1,172 @@
+        infoDialog.open();
+
+        // get the dial-in pin
+        String displayedPin = infoDialog.getPinNumber();
```
Should this method remove the "#" from the pin?

damencho replied:
Make sense, I will change it that way, it will not break other tests.

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/DialInAudioTest.java:
```diff
@@ -0,0 +1,172 @@
+                restURI.getPort(),
+                restURI.getScheme());
+
+            HttpGet httpget = new HttpGet(restURI);
```
For the sake of page objects, should this request logic be hidden behind a method on the participant or even in the infodialog? I was thinking the ideal is for the tests to read like english as much as possible and for test logic to be hidden away somewhere else.

damencho replied:
Well, this is very specific for the current test. Initially had even named the test with REST in the name...

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/DialInAudioTest.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    /**
+     * Executes the rest configured web call, replacing the dial-in pin the url
```
I think there might be typos in this comment.

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/pageobjects/web/WebRemoteParticipant.java:
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright @ 2018 8x8 Pty Ltd
```
2019?

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/base/Participant.java:
```diff
@@ -543,4 +543,23 @@ public void saveHtmlSource(File outputDir, String fileName)
+    /**
+     * @return a list of representation of the remote participants thumbnails.
+     */
+    public abstract List<RemoteParticipant<T>> getRemoteParticipants();
```
Should this method be renamed to include something about it being a thumbnail representation?

damencho replied:
Any ideas ... do we have something similar on the other page objects?

virtuacoplenny replied:
There really isn't a good existing precedent for interactions with remote participant thumbnails; from what I know MeetUIUtils is still used, which takes in an endpoint id. Existing page objects for the most part deal with UI outside of the filmstrip and not dynamically generated like the remote thumbnails.

I think it comes down to what you want the RemoteParticipant to represent. Do you want it to be more of an encompassing class that includes remote participant data, and maybe other helper methods related to the data, as well as DOM interactions? I'm feeling like right now you want it to instead represent the thumbnail itself, considering the javadoc.

virtuacoplenny replied:
Then change the javadoc not to say it's returning a representation of the thumbnail?

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/pageobjects/web/WebRemoteParticipant.java:
```diff
@@ -0,0 +1,96 @@
+    private void clickOnRemoteMenuLink(String linkClassname)
+    {
+        // Open the remote video menu
+        WebElement cntElem
```
What is cnt? I'd prefer this be named to what it is, without abbreviation, because I don't think cnt is a common abbreviation.

damencho replied:
Haha, no idea :) I just moved the code. I will change it with something more readable.

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/pageobjects/web/WebRemoteParticipant.java:
```diff
@@ -0,0 +1,96 @@
+        WebElement cntElem
+            = driver.findElement(By.id("participant_" + this.getEndpointId()));
+        String remoteVideoMenuButtonXPath
+            = TestUtils.getXPathStringForClassName(
```
Usage of this helper is off to me, considering this might be the only place it gets used, meaning it has inconsistent usage so I think we should use it more or kill it. You can probably kill this helper by changing what By you're using. Without changing jitsi-meet more, you can just look for By.classname (or maybe use an xpath but check it includes the class). Ideally you try to keep selectors as concise as possible so that when the dom/layout changes the tests don't have to be modified, and tying to a span adds one more reason to change.

damencho replied:
I will look at it, tomorrow to killing it.

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/pageobjects/web/WebRemoteParticipant.java:
```diff
@@ -0,0 +1,96 @@
+        MeetUIUtils.clickOnElement(driver,
+            "ul.popupmenu a." + linkClassname, true);
+
+        // wait for confirm muting to display
```
Probably need to update these comments because they're currently mute specific.

damencho replied:
Ups.

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/pageobjects/web/WebRemoteParticipant.java:
```diff
@@ -0,0 +1,96 @@
+
+        // confirm muting
+        MeetUIUtils.clickOnButton(driver,
+            "modal-dialog-ok-button", true);
```
Make a generic modal page object? Out of scope to do so?

damencho replied:
Will think of it ... I see we use for example modal-dialog-ok-button in 3 places at the moment...

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/web/WebParticipant.java:
```diff
@@ -65,6 +67,12 @@
+     *  The javascript code which returns an array of remote participant IDs.
+     */
+    public static final String GET_REMOTE_PARTICIPANT_IDS =
+        "return APP.conference._room.getParticipants().map(p => p._id);";
```
Do you think this is the best approach for the given desired changes? I'm generally not a fan of executing JS like this when it is avoidable, that way the end-to-end act more like a regular user. I think you can get the same info by scraping the dom, although then you rely on a stateful dom...so using JS seems okay too considering there's a lot of precedent for it already.

damencho replied:
Well, we are getting our own id using JS APP.conference ... so using the same way and getting remotes, I don't think is a problem.

paweldomas commented:
@damencho just replying to request for review from me: I think it's in good hands and it will be enough if Lenny looks at it, so don't wait for my review and just merge when he approves it.

damencho commented:
@virtuacoplenny Can you make another round of review. The changes for the modal dialog are in a separate commit and the rest of the changes addressing your comments are in the other commit. Thanks.

virtuacoplenny commented on src/test/java/org/jitsi/meet/test/pageobjects/web/ModalDialogHelper.java:
```diff
@@ -0,0 +1,63 @@
+            driver,
+            "//button[@id='" + OK_BUTTON +"']",
+            5);
+    }
```
TestUtils#click does this already.

damencho replied:
I just updated it.
